### PR TITLE
Fix treatment of MIR slices in overrides

### LINKED ISSUE #2045: SAW MIR override with slice argument fails to match
Here are two Rust functions, both of which take a slice argument:

```rs
// test.rs
pub fn f(s: &[u8]) -> u8 {
    s[0] + s[1]
}

pub fn g(s: &[u8]) -> u8 {
    f(s)
}
```

`g` does nothing but call `f`, so I would expect that the following compositional specification for `g` would verify:

```
// test.saw
enable_experimental;

let oneThroughFive = mir_term {{ [1, 2, 3, 4, 5] : [5][8] }};

let f_spec = do {
  a <- mir_alloc (mir_array 5 mir_u8);
  mir_points_to a oneThroughFive;

  mir_execute_func [mir_slice_value a];

  mir_return (mir_term {{ 3 : [8] }});
};

m <- mir_load_module "test.linked-mir.json";

f_ov <- mir_verify m "test::f" [] false f_spec z3;
mir_verify m "test::g" [f_ov] false f_spec z3;
```

While `f` does indeed verify against `f_spec`, `g` does not verify, much to my surprise:

```
$ ~/Software/saw-1.1/bin/saw test.saw
[13:23:34.144] Loading file "/home/ryanscott/Documents/Hacking/Haskell/saw-script/test.saw"
[13:23:34.209] Verifying test/8d8066da::f[0] ...
[13:23:34.224] Simulating test/8d8066da::f[0] ...
[13:23:34.225] Checking proof obligations test/8d8066da::f[0] ...
[13:23:34.241] Proof succeeded! test/8d8066da::f[0]
[13:23:34.280] Verifying test/8d8066da::g[0] ...
[13:23:34.294] Simulating test/8d8066da::g[0] ...
[13:23:34.295] Matching 1 overrides of  test/8d8066da::f[0] ...
[13:23:34.295] Stack trace:
"mir_verify" (/home/ryanscott/Documents/Hacking/Haskell/saw-script/test.saw:18:1-18:11)
Symbolic execution failed.
Abort due to assertion failure:
  test.rs:6:7: 6:8: error: in test/8d8066da::g[0]
  All overrides failed during structural matching:
  *  Name: test/8d8066da::f[0]
     Location: /home/ryanscott/Documents/Hacking/Haskell/saw-script/test.saw:17:9
     Argument types: 
     - &[u8]
     Return type: u8
     Arguments:
     - <slice>
     at /home/ryanscott/Documents/Hacking/Haskell/saw-script/test.saw:8:3
     could not match specified value with actual value:
       actual (simulator) value: 0x1:[8]
       specified value:          let { x@1 = Prelude.Vec 8 Prelude.Bool
                                       x@2 = Cryptol.PLiteralSeqBool
                                               (Cryptol.TCNum 8)
           }
        in [ Cryptol.ecNumber (Cryptol.TCNum 1) x@1 x@2
       , Cryptol.ecNumber (Cryptol.TCNum 2) x@1 x@2
       , Cryptol.ecNumber (Cryptol.TCNum 3) x@1 x@2
       , Cryptol.ecNumber (Cryptol.TCNum 4) x@1 x@2
       , Cryptol.ecNumber (Cryptol.TCNum 5) x@1 x@2 ] : [5][8]
       type of actual value:    u8
       type of specified value: [u8; 5]
```

The reason that SAW gives is rather bewildering, as the actual and specified values aren't even of the same type.

### PR DESCRIPTION
**Pull request by RyanGlScott**
As noted in #2045, the previous way that MIR slice reference valus were handled in overrides was horribly broken, as it was registering raw slice references (of type `*const T`) at type `&[T; N]`. We now take steps to turn the raw slice reference into an array reference before registering it at type `&[T; N]`. See the new `Note [Matching slices in overrides]` for the full story.

Fixes #2045.

### CODE REVIEW

File: src/SAWScript/Crucible/MIR/Override.hs [hunk 1]
```diff
@@ -1204,48 +1204,61 @@ matchArg opts sc cc cs prepost md actual expectedTy expected =
...
+             expectedSliceLen <- arrRefTyLen expectedArrRefTy
+             unless (expectedSliceLen == actualSliceLen) fail_
+             -- ...then match the reference values.
+             matchSlice expectedArrRefTy expectedArrRef
+           MirSetupSliceRange expectedArrRef expectedStart expectedEnd -> do
+             -- Check that the length of the range of the expected slice
+             -- reference value matches that of the actual slice reference
+             -- value...
+             expectedArrRefTy <- typeOfSetupValue cc tyenv nameEnv expectedArrRef
+             let expectedSliceLen = expectedEnd - expectedStart
```
Comment (sauclovian-g) [inline src/SAWScript/Crucible/MIR/Override.hs:1258]: Shouldn't this check that ```expectedEnd``` is less than (or equal to) ```arrRefTyLen expectedArrRefTy```, at least as an assertion? (One would expect not to make it this far if it's not true, but it seems like a good thing to crosscheck.)

File: src/SAWScript/Crucible/MIR/Override.hs [hunk 2]
```diff
@@ -1317,6 +1330,41 @@ matchArg opts sc cc cs prepost md actual expectedTy expected =
...
+Instead, we want to call `matchArg` on the *array reference value* associated
+with a slice, not the raw reference value itself. To do this, we take the raw
+reference value and use mirRef_peelIndexIO, a crucible-mir memory model
+operation which "peels back" the indexing operation that raw slice references
+use, thereby turning a `*const T` value into a `&[T; N]` value. It's a bit
+indirect, but it avoids needing to plumb around the original array reference
+value alongside the slice's raw reference value.
+
+This assumes that all slice reference values passed to an override were derived
+from crucible-mir's indexing operations, as this is crucible for
```
Comment (sauclovian-g) [inline src/SAWScript/Crucible/MIR/Override.hs:1362]: I think this got garbled

**Review by sauclovian-g — APPROVED** [review]
Some kibitzing, otherwise looks plausible (not that I really am yet in a position to notice if there's something subtle wrong)

**Comment by sauclovian-g** [comment]
(more importantly, it successfully addresses my problem)

File: src/SAWScript/Crucible/MIR/Override.hs (hunk 1 above)
Comment (RyanGlScott) [reply src/SAWScript/Crucible/MIR/Override.hs:1258]: A good question. In general, most of these style of checks are delegated to the code in `resolveSetupVal` (e.g., [here](https://github.com/GaloisInc/saw-script/blob/be24399f385acd966b1cb59331f636e8f1609047/src/SAWScript/Crucible/MIR/ResolveSetupValue.hs#L613-L620)), which is invoked on any value before it is passed as an argument to an override. That will ensure that a slice range value is consistent with its own underlying array's length. The checks here in `matchArg` are specifically for checking the values of two different slice values, which isn't checked as part of `resolveSetupVal`.

Admittedly, this separation of concerns isn't always adhered to in practice, as there are parts of `matchArg` that also do `resolveSetupVal`-style self-consistency checks. In the sake of doing the right thing in this specific example, I will:

1. Leave a comment here which states that the `expectedEnd <= arrRefTyLen expectedArrRefTy` check is done elsewhere, and
2. Add an expected-to-fail test case that ensures that this check actually occurs.

Comment (RyanGlScott) [reply src/SAWScript/Crucible/MIR/Override.hs:1258]: Actually, there are already test cases for (2) in [`test_mir_verify_slices`](https://github.com/GaloisInc/saw-script/blob/be24399f385acd966b1cb59331f636e8f1609047/intTests/test_mir_verify_slices/test.saw#L32-L52).

**Pull request merged**